---
title: "LFQ distribution of slim-domain interactions vs background"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: united
author: Bora Uyar
---

`r date()`

# Motivation

The goal of this analysis is to inspect the LFQ scores in the context of SLiM-Domain interactions without considering the SILAC values. 

Approach: From the proteomics screen, find out if LFQ values of
the interactions that can be explained by slim-domain pairs
are higher than the LFQ values of the background distribution.

- For each protein, I find the mean LFQ value of the protein with all peptides (including all genotypes: wt/phos/mut)
- Then I convert the LFQ value to a z-score (proteins with a NA value are converted to minimum LFQ value observed in the screen ~ 15.7). Thus, we can compare LFQ values across proteins. A high z-score means the protein preferentially binds to the target peptide. 
- Finally I check, if the protein-peptide interaction with high z-scores are enriched for slim-domain interactions. 

# Setup

Import required libraries
```{r setup}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, fig.height = 7, fig.width = 9)
suppressPackageStartupMessages(library(slimR))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(parallel))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(cowplot))

options(bitmapType='cairo')
data.table::setDTthreads(12)
ggplot2::theme_set(ggpubr::theme_pubclean())
datadir <- '../data'
outdir <- '..'

if(!dir.exists(datadir)) {stop(datadir, "doesn't exist")}
if(!dir.exists(outdir)) {stop(outdir, "doesn't exist")}

figureDir <- file.path(outdir, "figures")
tableDir <- file.path(outdir, "tables")
if(!dir.exists(figureDir)) {dir.create(figureDir)}
if(!dir.exists(tableDir)) {dir.create(tableDir)}
```

Importing and processing peptide info:
```{r}
peptides <- data.table::fread(file.path(datadir, 'PeptideCandidates1.txt'))
peptides$type <- 'wt'
peptides[grepl('_', peptides$Name)]$type <- 'mut'
peptides[grepl('phos', peptides$Name),]$type <- 'phos'
# remove the leading lower-case 'p' in the phosphorylated form of the sequence
#peptides$Peptide <- gsub("p([STY])", paste0("_phos_", "\\1"), peptides$Peptide)
peptides$Peptide <- gsub("p([STY])", "\\1", peptides$Peptide)
# add leading and trailing "XXX" to the peptide sequence to avoid matching terminal motifs
peptides$Peptide <- paste0('XXX',peptides$Peptide,'XXX')
# assign experiment group ids to make it compatible with Expgroup ids in array data
peptides$Expgroup <- ceiling(peptides$ID/3)*3-2
```

Get ELM classes
```{r elm}
elms <- readRDS(file.path(datadir, 'elm_classes.RDS'))
```

Get uniprot accessions and corresponding gene names from the fasta file of human uniprot sequences
```{r uniprot}
message('Importing fasta sequences')
fasta <- Biostrings::readAAStringSet(file.path(datadir, 'uniprot.9606.Nov_1_2021.fasta'))
uniprot2geneName <- unique(data.table('uniprotAccession' = sub("^sp\\|(.+?)\\|.+$", "\\1", names(fasta)),
                                      'geneName' = sub("^.+ GN=(.+?) .+$", "\\1", names(fasta)),
                                      stringsAsFactors = FALSE))
# simplify fasta headers of fasta data
names(fasta) <- sub("^sp\\|(.+?)\\|.*", "\\1", names(fasta))
```

Get LFQ data, process protein ids, cleanup, assign uniprot ids, get mean LFQ value per interaction from replicates
```{r lfq}
#import LFQ interaction results 
lfqData <- fread(file.path(datadir, 'LFQinteractions.txt'))
# simplify uniprot accession/gene name fields => remove unreviewed ids and keep only one per row
cl <- parallel::makeCluster(4)
parallel::clusterExport(cl, varlist = c('lfqData', 'uniprot2geneName'))
s <- pbapply::pbsapply(cl = cl, strsplit(lfqData$Protein.IDs, ";"), function(x) {
  # pick first match
  intersect(x, uniprot2geneName$uniprotAccession)[1]
})
lfqData$uniprotAccession <- s
parallel::stopCluster(cl)

lfqData <- lfqData[!is.na(uniprotAccession)]

lfqData <- melt.data.table(lfqData, id.vars = 'uniprotAccession', 
                measure.vars = grep('wt|mut|phos', colnames(lfqData), value = T))
lfqData$variable <- as.character(lfqData$variable)
lfqData$type <- sub("(wt|phos|mut).+", "\\1", lfqData$variable)
lfqData$Expgroup <- gsub("\\.[0-9]+$", "", gsub("(wt|phos|mut)", "", lfqData$variable))

# get mean LFQ value per interaction 
lfqData <- lfqData[,mean(value, na.rm = T), by = c('uniprotAccession', 'type', 'Expgroup')][!is.na(V1)]
```

Import PFAM annotations from Uniprot and ELM-PFAM interaction information
```{r elm2pfam}
message("importing PFAM domains")
uniprot2pfam <- readRDS(file.path(datadir, 'uniprot2pfam.RDS'))
uniprot2pfam <- uniprot2pfam[uniprot2pfam$seqnames %in% unique(lfqData$uniprotAccession),]
uniprot2pfam <- unique(subset(uniprot2pfam, select = c('seqnames', 'pfam_acc', 'pfam_name', 'clan')))
colnames(uniprot2pfam) <- c('uniprotAccession', 'PFAM', 'NAME', 'CLAN')
message("importing ELM -> PFAM associations")
# download a table of ELM classes and their cognate PFAM domains
elm2pfam <- readRDS(file.path(datadir, 'elm2pfam.RDS'))

# simplify uniprot2pfam table:
# remove proteins that don't exist in the screen
uniprot2pfam <- uniprot2pfam[uniprotAccession %in% lfqData$uniprotAccession]
```

# Matching SLiMs to PFAM domains

For each peptide, and for each protein in the screen, find which elm-interacting domains exist. 
```{r get_slim_domain_interactions}
# for each peptide, and for each protein in the screen, find which elm-interacting domains exist
slimDomainInteractions <- do.call(rbind, pbapply::pblapply(unique(peptides$Expgroup), function(e) { #for each experiment group
  dt <- do.call(rbind, lapply(c(unique(peptides$type)), function(g) { # for each genotype
    # slims in peptide
    seq <- peptides[Expgroup == e][type == g]$Peptide
    slims <-sort(unique(slimR::searchSLiMs(seq, elms)[['SLiM']]))
    # get list of proteins detected for the experiment group 
    prots <- unique(lfqData[Expgroup == e][type == g]$uniprotAccession)
    
    # find elm-interacting domains among the detected interactors
    dt <- droplevels(merge(uniprot2pfam[uniprotAccession %in% prots],
                           elm2pfam[ELM_identifier %in% slims], by.x = 'NAME', by.y = 'Interaction_Domain_Name'))
    # get a list of slim-domain pairs for each protein
    do.call(rbind, lapply(unique(dt$uniprotAccession), function(x) {
      data.table('uniprotAccession' = x,
                 'slim_domain_pairs' =  paste(paste(dt[uniprotAccession == x]$NAME,
                                                    dt[uniprotAccession == x]$ELM_identifier, sep = "->"),
                                              collapse = '; '),
                 'genotype' = paste0(g, '.slim_domain_pairs'),
                 'Expgroup' = e)
    }))
  }))
}))

slimDomainInteractions <- dcast.data.table(slimDomainInteractions,
                                           Expgroup + uniprotAccession ~ genotype,
                                           value.var = 'slim_domain_pairs')
slimDomainInteractions$Expgroup <- as.character(slimDomainInteractions$Expgroup)
```

# Normalize LFQ score across peptides

Melt the table into matrix form. Scaling each row, assigns z-scores to each protein (normalizing w.r.t other peptides)

```{r}
x <- dcast(lfqData, Expgroup + type ~ uniprotAccession, value.var = 'V1')
M <- as.matrix(data.frame(x[,-c(1,2)], row.names = paste(x$Expgroup, x$type, sep = '_')))
# replace NA values with value from first permille 
M[is.na(M)] <- as.numeric(quantile(lfqData$V1, 1:1000/1000)[1])
M <- t(scale(M)) # (each protein is normalised w.r.t other peptides)
write.table(M, file = file.path(tableDir, 
                                'LFQinteractions.scaled_by_peptide.tsv'),
            sep = '\t', quote = F)
```


## Combining LFQ data with slim-domain pairs

I obtain the scaled/normalized values of all pairs of peptides/proteins and check the 
distribution of z-scores for those categorized by whether the slim-domain interactions 
have higher z-scores overall. 

```{r}
lfqScaled <- data.table::melt(as.data.table(M, keep.rownames = T), measure.vars = colnames(M))
lfqScaled$Expgroup <- unlist(lapply(strsplit(as.character(lfqScaled$variable), "_"), function(a) a[1]))
lfqScaled$type <- unlist(lapply(strsplit(as.character(lfqScaled$variable), "_"), function(a) a[2]))
lfqScaled$variable <- NULL
colnames(lfqScaled)[1] <- 'uniprotAccession'

# see if interactions explained by slim-domain pairs
# have higher LFQ values
lfqMerged <- merge(lfqScaled, slimDomainInteractions,
           by = c('uniprotAccession', 'Expgroup'), all = T)
lfqMerged$paired = FALSE
lfqMerged[type == 'wt'][!is.na(wt.slim_domain_pairs)]$paired <- TRUE
lfqMerged[type == 'mut'][!is.na(mut.slim_domain_pairs)]$paired <- TRUE
lfqMerged[type == 'phos'][!is.na(phos.slim_domain_pairs)]$paired <- TRUE

write.table(lfqMerged[paired == TRUE], 
            file = file.path(tableDir, 'LFQscaled_slim_domain_interactions.tsv'), 
            sep = '\t', quote = F)

```

We can see the LFQ values for paired interactions are higher than the rest. However, this seems to be true for all genotypes (phos/wt/mut). This can be explained by the fact that slim-domain pairs we defined may not be phosphorylation dependent docking motifs. For example, a lot of slim-domain pairs are defined where the peptide has a "MOD" type motif (e.g. MOD_GSK3) and the interacting protein has a Pkinase domain. This is a valid ELM-PFAM pair, however, this doesn't mean the protein with the kinase domain will dock onto this peptide.

```{r}
ggplot(lfqMerged, aes(x = value)) + geom_density(aes(fill = paired), alpha = 0.4) + 
  facet_wrap(~ type, ncol = 1) + labs(x = 'LFQ z-score', fill = 'SLiM-Domain Interaction')
ggsave(filename = file.path(figureDir, 
                            'lfq_zscore_vs_slimdomain_interactions.pdf'), width = 10, height = 8)
```

Therefore, we categorize slim-domain pairs by if the peptide has a phosphorylation dependent docking/ligand/degron motif and the interaction partner has a cognate domain. If we focus on slim-domain pairs, where the peptide has a phosphorylation dependent docking/ligand/degron motif (excluding modification/targeting/cleavage motifs), then we see the normalized LFQ scores for phos peptides are higher than both wt/mut peptides. 

```{r}
pat <- "(\\(\\[[ST]+\\]\\)|\\([YTS]\\))"
e <- names(grep(pat, unlist(elms[grep('MOD|TRG|CLV', names(elms), invert = T)]), value = T))
lfqMerged$phosDependentMotifs <- unlist(lapply(strsplit(lfqMerged$phos.slim_domain_pairs, "; "), 
                                               function(x) sum(gsub("^.+->", "", x) %in% e)))

ggboxplot(lfqMerged[phosDependentMotifs > 0], x = 'type', y = 'value', color = 'type',
         add = 'jitter') + stat_compare_means(method.args = list('alternative' = 'greater'), 
                                               comparisons = list(c('phos', 'wt'),
                                                                  c('phos', 'mut'))) + 
  labs(title = 'LFQ score distributions of interactions explained by\nphospho-dependent-SLiM-Domain pairs',
       x = 'peptide type', y = 'LFQ score')
ggsave(filename = file.path(figureDir, 
                            'lfq_zscore_phos_vs_wt_vs_mut.doc_lig_deg.pdf'), 
       width = 5, height = 7)

```

If we also exclude the degron motifs, which require double phosphorylation, then we see an even stronger difference. 
```{r}
ggboxplot(lfqMerged[phosDependentMotifs > 0][grep('DEG', phos.slim_domain_pairs, invert = T)], x = 'type', y = 'value', color = 'type',
         add = 'jitter') + stat_compare_means(method.args = list('alternative' = 'greater'), 
                                               comparisons = list(c('phos', 'wt'),
                                                                  c('phos', 'mut'))) +
    labs(title = 'LFQ score distributions of interactions explained by\nphospho-dependent-SLiM-Domain pairs',
       x = 'peptide type', y = 'LFQ score')
ggsave(filename = file.path(figureDir, 
                            'lfq_zscore_phos_vs_wt_vs_mut.doc_lig.pdf'), width = 5, height = 7)
```

We also break the interactions into groups based on the detected domain types in the interaction partners and compare LFQ scores slim-domain-paired interactions with the background. 

```{r}
# get top pfam domains with most detected interaction partners in the screen 
# the domains must be known to be an elm-interacting domain and must be found in proteins detected in the screen. 
domains_paired_count <- sapply(unique(elm2pfam$Interaction_Domain_Name), function(d) {
  dt <- lfqMerged[uniprotAccession %in% uniprot2pfam[NAME == d]$uniprotAccession]
  sum(dt$paired)
})
# select those that are paired in at least 10 interactions 
domains <- names(sort(domains_paired_count, decreasing = T)[1:10]) #names(which(domains_paired_count > 10))

# find proteins in the whole screen that contain the domain
plots <- sapply(simplify = F, domains, function(d) {
  dt <- lfqMerged[uniprotAccession %in% uniprot2pfam[NAME == d]$uniprotAccession]
  # Check if the paired interactions are actually explainable
  # by both the presence of the domain in the interaction partner
  # and the presence of a cognate motif in the peptide. 
  # check if a motif-domain pair is defined for the correct peptide form 
  dt$specific_paired <- FALSE
  slims <- NA
  for (peptide_type in c('mut', 'phos', 'wt')) {
    # assigned domain-paired to TRUE
    slimfield <- paste0(peptide_type,".slim_domain_pairs")
    dt[type == peptide_type][grepl(paste0(d,"->"), get(slimfield))]$specific_paired <- TRUE
    # extract list of motifs in the peptides: 
    dt_sub <- dt[type == peptide_type][grepl(paste0(d,"->"), get(slimfield))]
    slims <- c(slims, unique(gsub("->", "", unlist(stringi::stri_extract_all(str = unlist(strsplit(dt_sub[[slimfield]], ";")), 
                                                                                            regex = "->.+")))))
  }
  slims <- unique(slims[!is.na(slims)])
  
  if(nrow(dt[specific_paired == TRUE]) > 10) {
      ggboxplot(dt, y = 'value', color = 'specific_paired', add = 'jitter', facet.by = 'type', nrow = 1) +
    labs(x = 'LFQ z-score', fill = 'SLiM-Domain Interaction') + 
    ggtitle(label = paste(d, "domain"), 
            subtitle = paste(strwrap(paste(slims, collapse = '; '), width = 40), collapse = '\n')) +
    geom_hline(yintercept = 0, linewidth = 2, alpha = 0.2) + 
    scale_color_brewer(type = 'qual', palette = 3)  
  } else {
    return(NULL)
  }
})

plots <- plots[!unlist(lapply(plots, is.null))]

p <- cowplot::plot_grid(plotlist = plots, ncol = 3, labels = 'AUTO')
ggsave(plot = p, filename = file.path(figureDir, "lfq_dist_by_top_domains.pdf"), 
       units = 'in', width = 12, height = 12)
# only plot 14-3-3, WW, SH3, and SH2 domains 
p <- cowplot::plot_grid(plotlist = plots[c('14-3-3', 'WW', 'SH2', 'SH3_1')], 
                        ncol = 2, labels = 'AUTO')
ggsave(plot = p, filename = file.path(figureDir, "lfq_dist_by_top_domains.subset.pdf"), 
       units = 'in', width = 10, height = 8)

```

# Summary

When I look into the LFQ values normalized across peptides, I see a higher z-score
for slim-domain paired protein-peptide interactions. However, this is true for 
all genotypes. I would expect to see that e.g. phosphopeptides have a higher z-score
compared to wild-type or mutant peptides. Following up on this, we decided to 
categorize such slim-domain pairs by if the interaction is expected to be a phosphorylation-dependent docking interaction (excluding targeting/modification motifs). Then we observe that the 
phos peptides have higher scores than both wt/mut peptides. An interesting group are the degrons, which usually require double phosphorylation sites. The difference between phos and wt/mut get 
even bigger if we exclude degrons. This could be because degrons require double phosphorylation sites, while the peptides in the screen were designed with only a single phosphorylation. 

# Session Info
```{r}
print(sessionInfo())
```

