---
title: "Phospho-dependent domain discovery from the array data"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: united
author: Bora Uyar
---

`r date()`


Import required libraries
```{r setup}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, fig.height = 7, fig.width = 9)
suppressPackageStartupMessages(library(slimR))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(parallel))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(cowplot))

ggplot2::theme_set(ggpubr::theme_pubclean())
data.table::setDTthreads(12)
```


# Motivation

We'd like to find out if we can detect an enrichment of phospho-dependent binding domains in the array data.

# Data preparation

Importing and processing peptide info:
```{r }
datadir <- '../data/'
peptides <- data.table::fread(file.path(datadir,  'PeptideCandidates1.txt'))
peptides$type <- 'wt'
peptides[grepl('_', peptides$Name)]$type <- 'mut'
peptides[grepl('phos', peptides$Name),]$type <- 'phos'
# remove the leading lower-case 'p' in the phosphorylated form of the sequence
#peptides$Peptide <- gsub("p([STY])", paste0("_phos_", "\\1"), peptides$Peptide)
peptides$Peptide <- gsub("p([STY])", "\\1", peptides$Peptide)
# add leading and trailing "XXX" to the peptide sequence to avoid matching terminal motifs
peptides$Peptide <- paste0('XXX',peptides$Peptide,'XXX')
# assign experiment group ids to make it compatible with Expgroup ids in array data
peptides$Expgroup <- ceiling(peptides$ID/3)*3-2
```

Get uniprot accessions and corresponding gene names from the fasta file of human uniprot sequences
```{r uniprot}
message('Importing fasta sequences')
fasta <- Biostrings::readAAStringSet(file.path(datadir, 'uniprot.9606.Nov_1_2021.fasta'))
uniprot2geneName <- unique(data.table('uniprotAccession' = sub("^sp\\|(.+?)\\|.+$", "\\1", names(fasta)),
                                      'geneName' = sub("^.+ GN=(.+?) .+$", "\\1", names(fasta)),
                                      stringsAsFactors = FALSE))
# simplify fasta headers of fasta data
names(fasta) <- sub("^sp\\|(.+?)\\|.*", "\\1", names(fasta))
```

Get LFQ data, process protein ids, cleanup, assign uniprot ids, get mean LFQ value per interaction from replicates
```{r lfq}
#import LFQ interaction results 
lfqData <- fread(file.path(datadir, 'LFQinteractions.txt'))
# simplify uniprot accession/gene name fields => remove unreviewed ids and keep only one per row
cl <- parallel::makeCluster(10)
parallel::clusterExport(cl, varlist = c('lfqData', 'uniprot2geneName'))
s <- pbapply::pbsapply(cl = cl, strsplit(lfqData$Protein.IDs, ";"), function(x) {
  # pick first match
  intersect(x, uniprot2geneName$uniprotAccession)[1]
})
lfqData$uniprotAccession <- s
parallel::stopCluster(cl)

lfqData <- lfqData[!is.na(uniprotAccession)]

lfqData <- melt.data.table(lfqData, id.vars = 'uniprotAccession', 
                measure.vars = grep('wt|mut|phos', colnames(lfqData), value = T))
lfqData$variable <- as.character(lfqData$variable)
lfqData$type <- sub("(wt|phos|mut).+", "\\1", lfqData$variable)
lfqData$Expgroup <- gsub("\\.[0-9]+$", "", gsub("(wt|phos|mut)", "", lfqData$variable))

# get mean LFQ value per interaction 
lfqData <- lfqData[,mean(value, na.rm = T), by = c('uniprotAccession', 'type', 'Expgroup')][!is.na(V1)]

# normalize lfqData
x <- dcast(lfqData, Expgroup + type ~ uniprotAccession, value.var = 'V1')
M <- as.matrix(data.frame(x[,-c(1,2)], row.names = paste(x$Expgroup, x$type, sep = '_')))
M[is.na(M)] <- 15.7
lfqNorm <- t(scale(M))
```

Import PFAM annotations from Uniprot and download ELM-PFAM interaction information
```{r elm2pfam}
message("importing PFAM domains")
uniprot2pfam <- readRDS(file.path(datadir, 'uniprot2pfam.RDS'))
uniprot2pfam <- uniprot2pfam[uniprot2pfam$seqnames %in% unique(lfqData$uniprotAccession),]
uniprot2pfam <- unique(subset(uniprot2pfam, select = c('seqnames', 'pfam_acc', 'pfam_name', 'clan')))
colnames(uniprot2pfam) <- c('uniprotAccession', 'PFAM', 'NAME', 'CLAN')
message("importing ELM -> PFAM associations")
# download a table of ELM classes and their cognate PFAM domains
elm2pfam <- readRDS(file.path(datadir, 'elm2pfam.RDS'))

# simplify uniprot2pfam table:
# remove proteins that don't exist in the screen
uniprot2pfam <- uniprot2pfam[uniprotAccession %in% lfqData$uniprotAccession]
```


Get list of known phospho-dependent slim-domain pairs
```{r}
elms <- readRDS(file.path(datadir, 'elm_classes.RDS'))
pat <- "(\\(\\[[ST]+\\]\\)|\\([YTS]\\))"
phos_elms <- names(grep(pat, unlist(elms[grep('MOD|TRG|CLV', names(elms), invert = T)]), value = T))
phos_domains <- unique(elm2pfam[ELM_identifier %in% phos_elms]$Interaction_Domain_Name)
```

```{r}
# create feature matrix (features are protein domains and values are 1/0 if the protein has the corresponding domain)
domains <- unique(uniprot2pfam[uniprotAccession %in% rownames(lfqNorm)]$NAME)
M <- matrix(rep(0, nrow(lfqNorm) * length(domains)), nrow = nrow(lfqNorm), ncol = length(domains),
            dimnames = list(rownames(lfqNorm), domains))
# for each protein, if the protein has the domain, switch value to 1
dt <- droplevels(uniprot2pfam[uniprotAccession %in% rownames(lfqNorm)][!is.na(uniprotAccession)])
dtl <- lapply(split(dt, dt$uniprotAccession), function(x) unique(x$NAME))

counts <- do.call(rbind, lapply(rownames(M), function(p) {
  x <- M[p,]
  if(p %in% names(dtl)) {
      x[dtl[[p]]] <- 1
  }
  return(x)
}))
rownames(counts) <- rownames(M)
phos_domains <- intersect(colnames(counts), phos_domains)

```

Import Array Data (LFQ + SILAC dataset)

```{r}
arrayDataFile <- file.path(datadir, 'SILACLFQ_extended_loose_p001FD_1.txt')
arrayData <- data.table::fread(arrayDataFile)
# simplify uniprot accession/gene name fields => remove unreviewed ids and keep only one per row
cl <- parallel::makeCluster(10)
parallel::clusterExport(cl, varlist = c('arrayData', 'uniprot2geneName'))
s <- pbapply::pbsapply(cl = cl, strsplit(arrayData$Majority.protein.IDs, ";"), function(x) {
  # pick first match
  intersect(x, uniprot2geneName$uniprotAccession)[1]
})
arrayData$uniprotAccession <- s
parallel::stopCluster(cl)

# modify arrayData to have one line per protein & peptide in the screen
arrayData$SILACGroup <- gsub("[0-9]+$", "", arrayData$SILACGroup)
colnames(arrayData) <- gsub("_", ".", gsub("LFQsignificant", "lfq_", colnames(arrayData)))

arrayData <- data.table::dcast.data.table(arrayData[!is.na(uniprotAccession)],
                 Expgroup + uniprotAccession + lfq.Wt.loose + lfq.Phos.loose + lfq.Mut.loose ~ SILACGroup,
                 value.var = 'Median.SILAC.ratio')
# keep interactions that are not NA for any silac values
to_keep <- apply(arrayData[,c('phos_mut', 'wt_mut', 'wt_phos')], 1, function(x) sum(!is.na(x)) == 3)
arrayData <- arrayData[to_keep]
```

# Domain Enrichment Analysis 

The first approach is to do a hypergeometric test to see which domains are enriched among proteins that preferentially bind phosphorylated forms of the peptides compared to the background (universal set of proteins detected in the screen).

Find proteins that are LFQ+ for phosphorylated form of the peptide. 
For each domain, count the number of proteins containing the domain among this population and compare to the background distribution. 


```{r}
t <- colSums(counts[,phos_domains])
p1 <- ggplot(data.table('domain' = names(t), 'count' = as.numeric(t)),
        aes(x = reorder(domain, -count), y = count)) + geom_bar(aes(fill = log(count)), 
                                                                stat = 'identity', show.legend = F) + 
  labs(x = 'Phospho-domain', y = 'Number of proteins in the screen') + geom_hline(yintercept = 5)
```

```{r}
candidates <- unique(arrayData[lfq.Phos.loose == '+']$uniprotAccession)
# only considering domains that are found in at least 5 proteins in the screen 
stats <- do.call(rbind, pbapply::pblapply(names(which(colSums(counts) > 5)), function(d) {
  # find the number of proteins in the screen containing the domain 
  bg_count <- sum(counts[,d]) #number of proteins in the screen containing the domain
  phos_count <- sum(counts[candidates, d]) # number of proteins that are LFQ+ for phos peptide containing the domain 
  exp_count <- length(candidates) * bg_count/nrow(counts) # number of proteins expected to observe among LFQ+ phos 
  t <- binom.test(x = phos_count, n = length(candidates), p = bg_count / nrow(counts), alternative = 'greater') # p.value for hypergeometric test 
  data.table('domain' = d, 'phos' = phos_count, 'expected' = exp_count, 'pval' = t$p.value)
}))[order(pval)]


p2 <- ggplot(stats[pval < 0.05], 
        aes(x = reorder(domain, pval), y = -log10(pval))) + 
  geom_bar(aes(fill = domain %in% phos_domains), 
               stat = 'identity', show.legend = F) + coord_flip() + 
  labs(x = 'Domain', y = '-log10(pval) (binomial test)') + 
  scale_fill_brewer(type = 'qual', palette = 3) + 
  geom_hline(yintercept = -log10(0.05))

```


# using silac ratios to find candidates 

With this approach, we rank the domains by the average silac ratios of lfq phos + interaction partners across peptides 

```{r}

# phos vs mut 
phos_mut_dt <- arrayData[lfq.Phos.loose == '+', mean(phos_mut), by = 'uniprotAccession']

# for each domain find proteins that contain the domain, get the mean silac ratios  
stats_silac <- do.call(rbind, pbapply::pblapply(colnames(counts), function(d) {
  selected <- names(which(counts[,d] == 1))
  data.table::data.table('domain' = d, 'mean_phos_mut' = mean(phos_mut_dt[uniprotAccession %in% selected]$V1, na.rm = T))
}))

```

Top-ranking domains in terms of the mean SILAC ratios phos/mut across peptides 

```{r}
p3 <- ggplot(stats_silac[order(mean_phos_mut, decreasing = T)][1:30], 
             aes(x = reorder(domain, mean_phos_mut), y = mean_phos_mut)) + 
  geom_bar(stat = 'identity', aes(fill = domain %in% phos_domains), show.legend = T) + coord_flip() + 
  labs(x = 'Domain', y = 'Mean SILAC ratio (phos/mut) across peptides') + 
  scale_fill_brewer(type = 'qual', palette = 3) + 
  theme(legend.position = 'bottom')
```


```{r, fig.width=10, fig.height=7}
cowplot::plot_grid(
  cowplot::plot_grid(p1, p2, ncol = 1, labels = c('A', 'B')), 
  p3, ncol = 2, labels = c('', 'C') 
)
ggsave(filename = 'phospho_domain_analysis.pdf', width = 12, height = 8)
```

- Only few SLiM-binding phospho-dependent domains are found in at least 5 proteins in the whole screen. (Panel A)
- Among these 14-3-3 and SH2 domains are enriched (compared to the domain distribution of the full set of proteins detected in the screen) for interaction partners which are LFQ+ for the phosphorylated form of the peptides. (Panel B)
- If we look into the average phos/mut SILAC ratios for the proteins which are LFQ+ for phoshorylated form of the peptides, we can see the 14-3-3, SH2, and WW make it among the top 30 domains. Other domains are potentially those that are likely to co-occur with phospho-dependent domains. But, this could serve as a list of potential phospho-recognition domains based on this screen. (Panel C)

# Session Info
```{r}
print(sessionInfo())
```

